lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -832,5 +894,25 @@ public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
       return new DefaultBulkScorer(new ConstantScoreScorer(this, 0f, ScoreMode.COMPLETE_NO_SCORES, disi));
     }
 
+    // Perform a cache load asynchronously
+    private void cacheAsynchronously(LeafReaderContext context, IndexReader.CacheHelper cacheHelper) throws IOException {
+      FutureTask<Void> task = new FutureTask<>(() -> {
+        DocIdSet localDocIdSet = cache(context);
+        putIfAbsent(in.getQuery(), localDocIdSet, cacheHelper);
+
+        //remove the key from inflight -- the key is loaded now
+        inFlightAsyncLoadQueries.remove(in.getQuery());
+        return null;
+      });
+      /*
+       * If the current query is already being asynchronously cached,
+       * do not trigger another cache operation
+       */
+      if (inFlightAsyncLoadQueries.contains(in.getQuery())) {
+        return;
+      }
+      inFlightAsyncLoadQueries.add(in.getQuery());
```

Comment:
Hmm still a concurrency bug here, allowing two threads to cache the same query -- instead of separate `.contains` and `.add` calls, can you simply call `.add` and look at its return value to decide whether you should in fact cache?

Reply:
I was hesitant to do that since that would entail moving out the addition of the entry in the map from the actual triggering of cache load. However, there is minimal risk involved in decoupling these two, and will indeed safeguard us completely from two threads caching the same query. Fixed and updated